File: source/python/e2c/parser.py
#
# Copyright 2017 The E2C Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

from typing import \
    Dict, \
    List, \
    Callable

from . import const
from . import errors
from .actor import Actor


class Parser(object):
    """
    Represents a class to parse the graph and build relations between them.
    """

    def __init__(self, actors: Dict[str, Actor], create_actor: Callable[[str], Actor]):
        """
        Represents a class to parse the graph and build relations between them.

        :type actors: :class:`Dict[str, Actor]`
        :param actors: The list to add the handeled actors.

        :type create_actor: type: `Callable[[str], Actor]`
        :param create_actor: The function to build new actors.
        """
        self._actors = actors
        self._create_actor = create_actor

    def run(self, script: List[str], out_name: Callable[[str], None]):
        """
        Starts the parsing.

        :type script: List[str]
        :param script: The script to parse.

        :type out_name: Callable[[str], Actor]
        :param out_name: The function to receive the name of graph.

        :rtype: None
        """
        if not ''.join(script):
            raise errors.E2CParserError('No data to parse!')

        for index, line in enumerate(script, 1):
            line = line.replace('\n', '').replace(' ', '').strip()

            pos = line.find(const.COMMENT)
            if pos >= 0:
                line = line[:pos] if line else None
            if not line:
                continue

            if line.startswith('[') and line.endswith(']'):
                out_name(line[1:-1])
                continue

            if const.EDGE not in line:
                raise errors.E2CParserError(
                    'Missing {} in line {}!'.format(const.EDGE, index))

            left_actor_name_and_param, right_actor_name = line.split(const.EDGE)
            if not right_actor_name:
                raise errors.E2CParserError(
                    'Missing actor in line {}!'.format(index))

            left_actor_name, left_param = left_actor_name_and_param.split('.', 1)
            left_actor_name = left_actor_name or const.SELF

            if left_actor_name not in self._actors:
                self._actors[left_actor_name] = \
                    self._create_actor(left_actor_name)

            if right_actor_name not in self._actors:
                self._actors[right_actor_name] = \
                    self._create_actor(right_actor_name)

            self._actors[left_actor_name].on(
                left_param, self._actors[right_actor_name])

File: source/dotnet/e2c/Parser.cs
//
// Copyright 2017 The E2C Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

using System;
using System.Collections.Generic;

namespace E2c
{
    /// <summary>
    /// Represents a class to parse the graph and build relations between them.
    /// </summary>
    public class Parser
    {
        private Dictionary<string, Actor> actors;
        private Func<string, Actor> createActor;

        /// <summary>
        /// Represents a class to parse the graph and build relations between them.
        /// </summary>
        /// <param name="actors">The list to add the handeled actors.</param>
        /// <param name="createActor">The function to build new actors.</param>
        public Parser(Dictionary<string, Actor> actors, Func<string, Actor> createActor)
        {
            this.actors = actors;
            this.createActor = createActor;
        }

        /// <summary>
        /// Starts the parsing.
        /// </summary>
        /// <param name="script">The script to parse.</param>
        /// <param name="outName">The function to receive the name of graph.</param>
        public void Run(string[] script, Func<object, string> outName)
        {           
            if (String.Join("", script).Trim() == String.Empty)
                throw new E2CParserError("No data to parse!");

            int index = 0;
            foreach (string nextLine in script)
            {
                index++;
                string line = nextLine.Replace(
                    "\n", String.Empty).Replace(
                        " ", String.Empty).Trim();

                if (string.IsNullOrEmpty(line))
                    continue;

                int pos = line.IndexOf(Const.COMMENT);
                if (pos >= 0)
                {
                    line = line.Substring(0, pos);
                    if (String.IsNullOrEmpty(line))
                        continue;
                }

                if (line.StartsWith("[") && line.EndsWith("]"))
                {
                    outName(line.Substring(1, line.Length - 2));
                    continue;
                }

                if (!line.Contains(Const.EDGE))
                    throw new E2CParserError(
                        String.Format(
                            "Missing {0} in line {1}!", Const.EDGE, index));

                var (leftActorNameAndParam, rightActorName) = this.Split(line, Const.EDGE);
                if (String.IsNullOrEmpty(rightActorName))
                    throw new E2CParserError(
                        String.Format(
                            "Missing actor in line {0}!", index));

                var (leftActorName, leftParam) = this.Split(leftActorNameAndParam, ".");
                if (String.IsNullOrEmpty(leftActorName))
                    leftActorName = Const.SELF;

                if (!this.actors.ContainsKey(leftActorName))
                    this.actors.Add(leftActorName, this.createActor(leftActorName));
                if (!this.actors.ContainsKey(rightActorName))
                    this.actors.Add(rightActorName, this.createActor(rightActorName));

                this.actors[leftActorName].On(
                    leftParam, this.actors[rightActorName]);
            }

        }

        /// <summary>
        /// Represents the helper method to split the specified 
        /// line by given seperator.
        /// </summary>
        /// <param name="line"></param>
        /// <param name="seperator"></param>
        /// <returns></returns>
        private Tuple<string, string> Split(string line, string seperator)
        {
            string[] result = line.Split(new String[]{seperator}, StringSplitOptions.None);
            return new Tuple<string, string>(result[0], result[1]);
        }

    }
}